File: MInimum time to fulfil all orders.cpp
class Solution{
    public:
  
    // Function to return number of donuts prepared by R rank in T time
    int countDonut(int R, int T)
    {
        int cnt = 1;
        
        while(T-R*cnt >= 0)
        {
            T -= R*cnt;
            cnt++;
        }
        
        return cnt-1;
    }
    
    int findMinTime(int N, vector<int>&A, int L){
        //write your code here
        int ans = INT_MAX;
        int low = 0, high = INT_MAX;
        
        while(low <= high)
        {
            int mid = low + (high-low)/2;
            int cnt = 0;
            
            for(auto it : A)
            cnt += countDonut(it, mid);
            
            if(cnt >= N)
            {
                ans = mid;
                high = mid-1;
            }
            else
                low = mid+1;
        }
        
        return ans;
    }
};

File: Minimum time to fulfil all orders.java
class Solution {
    public static int findMinTime(int n, int x, int[] arr) {
        // code here
        Arrays.sort(arr);
        int l=arr[0];
        int h=arr[x-1]*n*(n+1)/2;
        int time=0;
        while(l<=h){
            int mid=(l+h)/2;
            if(isPossible(mid,n,arr)){
                time=mid;
                h=mid-1;
            }else{
                l=mid+1;
            }
        }
        return time;
    }
    static boolean isPossible(int mid,int n,int[]arr){
        int donots=0;
        for(int t:arr){
            int timeTaken=0;
            int timeWillTake=t;
            while(timeTaken+timeWillTake<=mid){
                donots++;
                timeTaken+=timeWillTake;
                timeWillTake+=t;
            }
        }
        return donots>=n;
    }
}
        
